text/0000-local-loop-bindings.md

```diff
@@ -0,0 +1,218 @@
+
+```rust
+loop i = 1 {
+    if i <= 10 {
+        println!("iteration {}", i);
+        continue i + 1;
+    }
+    break;
+}
+```
+
+Since the end of the loop is never reached, the return value is not required to be the type of the binding, here.
+
+A loop without bindings (`loop { /* body */ }`) will be the same as this:
+
+```rust
+loop () = () {
+    /* body */
+}
+```
+
+This will not be a breaking change, since it's not allowed to have values other than `()` from a loop.
+
+A simple example from the book looks like this:
+
+```rust
+let mut x = 5;
+let mut done = false;
+
+while !done {
+    x += x - 3;
+
+    println!("{}", x);
+
+    if x % 5 == 0 {
+        done = true;
+    }
+}
+```
+
+Using the new syntax, this could be rewritten as this:
+
+```rust
+loop (mut x, done) = (5, false) {
+    if done {
+        break;
+    }
+    x += x - 3;
+
+    println!("{}", x);
+
+    if x % 5 == 0 {
+        (x, true)
+    } else {
+        (x, false)
+    }
+}
+```
+
+This is, how you would define factorial using a loop now:
+
+```rust
+fn factorial(x: i32) -> i32 {
+    loop (result, count) = (1, x) {
+        if count == 1 {
+            break result;
+        }
+        (result * count, count - 1)
+    }
+}
+```
+
+With explicit `continue`, it can look like this:
+
+```rust
+fn factorial(x: i32) -> i32 {
+    loop (result, count) = (1, x) {
+        if count == 1 {
+            break result;
+        } else {
+            continue (result * count, count - 1);
+        }
+    }
+}
+```
+
+Using `break` here allows copying code without having to modify it, when not using a specific function.
+
+Labels will also work. When using `continue` with a label, the arguments to continue must match the loop binding signature connected to the label, in case the label is connected with a loop.
+
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+The syntax extension should not cause any issues with backwards compatability.
+
+It's just an extended syntax for `loop` in a place, where currently nothing is allowed yet.
+
+The expansion of the new syntax will be shown for an example.
+
+New syntax:
+
+```rust
+loop (a, mut b) = (x, y) {
+    /* body */
+}
+```
+
+Current syntax:
+
+```rust
+{ // ensure global bindings to be inaccessible after the loop
+    let mut binding = (x, y);
+    loop {
+        let (a, mut b) = binding;
+        binding = {
+            /* body */
+        }
+    }
+}
+```
+
+This expansion should cover the common case.
+
+A `continue value` in the body would expand to `binding = value; continue;
+
+Internally there may be more efficient ways to implement this.
+
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+This adds more options to the language, which also makes the language more complicated, but it should be pretty intuitive, how it works.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
```

Comment:
Maybe this is bikeshed, but personally, I would find it good to reuse `let pat = expr` so that you write:

```rust
loop let (mut x, done) = (5, false) {
    if done { break; }

    x += x - 3;
    println!("{}", x);
    (x, x % 5 == 0)
}
```

Reply:
I thought, `while let` and `if let` are extended versions of `if` and `while`, which take bools.
Here we don't have arguments anyway, so `let` is not necessary.
But when having a `let`, would it be better to accept refutable patterns, too, for consistency with `while let` and `if let`?

Also when thinking about adding more possible keywords after `while` and `if` not having a `let` only after `loop`, but also having the same new keywords (maybe `while const`, `if const` and `loop const`) will be inconsistent. 

So yeah, `loop let` is the syntax, I'd prefer, too.

Reply:
> I thought, `while let` and `if let` are extended versions of `if` and `while`, which take bools.
> Here we don't have arguments anyway, so `let` is not necessary.

I don't think it's for necessity; it's clearly not necessary syntactically; but it helps with clarity of highlighting that a pattern comes next. Just by seeing `let p = q` I know that `p` is a pattern and `q` is an expression. This reduces the overall complexity of adding `loop let`. With `loop p = q { .. }`  that isn't as clear and complexity costs increase.

> But when having a `let`, would it be better to accept refutable patterns, too, for consistency with `while let` and `if let`?

Sure 👍 

> Also when thinking about adding more possible keywords after `while` and `if` not having a `let` only after `loop`, but also having the same new keywords (maybe `while const`, `if const` and `loop const`) will be inconsistent.

Not sure what `while const` would be...

> 
> So yeah, `loop let` is the syntax, I'd prefer, too.

❤️